python/cudf/cudf/core/dataframe.py

```diff
@@ -1906,18 +1928,27 @@ def rename(self, mapper=None, columns=None, copy=True, inplace=False):
         Notes
         -----
         Difference from pandas:
-          * Support axis='columns' only.
-          * Not supporting: index, level
+            * Not supporting: level
 
         Rename will not overwite column names. If a list with duplicates is
         passed, column names will be postfixed with a number.
         """
-        # Pandas defaults to using columns over mapper
-        if columns:
-            mapper = columns
+        if mapper is None and index is None and columns is None:
+            return self.copy(deep=copy)
 
-        out = DataFrame(index=self.index)
-        if isinstance(mapper, Mapping):
+        index = mapper if index is None and axis in (0, "index") else index
+        columns = (
+            mapper if columns is None and axis in (1, "columns") else columns
+        )
+
+        if index:
+            out = DataFrame(
+                index=[index.get(item, item) for item in self.index]
```

Comment:
You can't loop over a DataFrame index. That copies device --> host element by element.